### ISSUE TITLE
BUG: Vite failed to resolve entry for package `@octokit/plugin-throttling`. The package may have incorrect main/module/exports specified in its package.json

### ISSUE THREAD

**Issue by HeavenlyEntity** [octokit/plugin-throttling.js#586]
### Specs
- Version: `5.1.1`
- Node Version: `18.12.0`
- Vite Version: `4.2.1`

### Expected Behavior

Initialize `new Octokit()` and access the object of functions. On my dev server.

### Actual Behavior

When trying the initialize the `new Octokit()` constructor class it fatally errors my application and the dev build fails. Then I get this lovely error. 

```shell
 Failed to resolve entry for package "@octokit/plugin-throttling". The package may have incorrect main/module/exports specified in its package.json. [plugin vite:dep-pre-bundle]

    node_modules/esbuild/lib/main.js:1360:21:
      1360 │         let result = await callback({
           ╵                      ^

    at packageEntryFailure (file:///C:/Users/user/Documents/GitHub/test-app/node_modules/vite/dist/node/chunks/dep-79892de8.js:23266:11)
    at resolvePackageEntry (file:///C:/Users/user/Documents/GitHub/test-app/node_modules/vite/dist/node/chunks/dep-79892de8.js:23263:5)
    at tryNodeResolve (file:///C:/Users/user/Documents/GitHub/test-app/node_modules/vite/dist/node/chunks/dep-79892de8.js:23000:20)
    at Context.resolveId (file:///C:/Users/user/Documents/GitHub/testapp/node_modules/vite/dist/node/chunks/dep-79892de8.js:22762:28)
    at Object.resolveId (file:///C:/Users/user/Documents/GitHub/test-app/node_modules/vite/dist/node/chunks/dep-79892de8.js:43314:46)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async file:///C:/Users/user/Documents/GitHub/test-app/node_modules/vite/dist/node/chunks/dep-79892de8.js:63611:21
    at async file:///C:/Users/user/Documents/GitHub/test-app/node_modules/vite/dist/node/chunks/dep-79892de8.js:23589:34
    at async requestCallbacks.on-resolve (C:\Users\user\Documents\GitHub\test-app\node_modules\esbuild\lib\main.js:1360:22)
    at async handleRequest (C:\Users\user\Documents\GitHub\test-app\node_modules\esbuild\lib\main.js:723:13)

  This error came from the "onResolve" callback registered here:

    node_modules/esbuild/lib/main.js:1279:20:
      1279 │       let promise = setup({
           ╵                     ^

    at setup (file:///C:/Users/user/Documents/GitHub/test-app/node_modules/vite/dist/node/chunks/dep-79892de8.js:23569:19)
    at handlePlugins (C:\Users\user\Documents\GitHub\test-app\node_modules\esbuild\lib\main.js:1279:21)
    at buildOrContextImpl (C:\Users\user\Documents\GitHub\test-app\node_modules\esbuild\lib\main.js:968:5)
    at Object.buildOrContext (C:\Users\user\Documents\GitHub\test-app\node_modules\esbuild\lib\main.js:776:5)
    at C:\Users\user\Documents\GitHub\test-app\node_modules\esbuild\lib\main.js:2172:68
    at new Promise (<anonymous>)
    at Object.context (C:\Users\user\Documents\GitHub\test-app\node_modules\esbuild\lib\main.js:2172:27)
    at Object.context (C:\Users\apmausers\Documents\GitHub\test-app\node_modules\esbuild\lib\main.js:2012:58)
    at prepareEsbuildOptimizerRun (file:///C:/Users/user/Documents/GitHub/test-app/node_modules/vite/dist/node/chunks/dep-79892de8.js:44840:35)

  The plugin "vite:dep-pre-bundle" was triggered by this import

    node_modules/octokit/dist-web/index.js:5:27:
      5 │ import { throttling } from '@octokit/plugin-throttling';

```

### Note
I updated to the latest version after the `package.json` fix hoping to fix this but vite is not liking something here.

**Comment by wolfy1339** [comment]
The issue with `package.json` that was fixed was files not getting published to npm.

There is an issue with `main`, `module` fields in `package.json`.

Currently `main` points to the CommonJS bundle for NodeJS, and the `module` field points to the web bundle.

Can you try patching the `package.json` in your `node_modules` lie in this commit https://github.com/octokit/webhooks.js/pull/792/commits/ffbf0ac70ae8fb901029a169b1c677e6a55bef7d and see if it works?

**Comment by HeavenlyEntity** [comment]
> Can you try patching the package.json in your node_modules like in this commit https://github.com/octokit/webhooks.js/commit/ffbf0ac70ae8fb901029a169b1c677e6a55bef7d and see if it works?

Yeah @wolfy1339 I can try that. Yet I'm afraid this will break in a production setting after I build the app. But correct me if I'm wrong here will it be fine after the app is built?

**Comment by wolfy1339** [comment]
The build should have no affect to the `package.json` in the `node_modules`. It will only change once you update/change dependencies

**Comment by HeavenlyEntity** [comment]
@wolfy1339 Its a no go. After doing that I get a new error:

```shell
Uncaught TypeError: Cannot read properties of undefined (reading 'prototype')
    at node_modules/@octokit/request/node_modules/node-fetch/lib/index.mjs (index.mjs:1165:65)
    at __init (chunk-AC2VUBZ6.js?v=6002d3fa:8:56)
    at node_modules/@octokit/request/dist-web/index.js (index.js:5:14)
    at __init (chunk-AC2VUBZ6.js?v=6002d3fa:8:56)
    at node_modules/@octokit/core/dist-web/index.js (index.js:7:14)
    at __init (chunk-AC2VUBZ6.js?v=6002d3fa:8:56)
    at @octokit_core.js?v=6002d3fa:7:1
```

**Comment by wolfy1339** [comment]
That's a lot better. At least it's getting somewhere farther. The error seems related to `node-fetch`
Can you apply the same patch to the other octokit modules?

**Comment by HeavenlyEntity** [comment]
> The error seems related to node-fetch
Can you apply the same patch to the other octokit modules?

I have done that now its still breaking on this:

```shell
 Uncaught TypeError: Cannot read properties of undefined (reading 'prototype')
    at node_modules/@octokit/request/node_modules/node-fetch/lib/index.mjs (index.mjs:1165:65)
    at __init (chunk-AC2VUBZ6.js?v=6002d3fa:8:56)
    at node_modules/@octokit/request/dist-web/index.js (index.js:5:14)
    at __init (chunk-AC2VUBZ6.js?v=6002d3fa:8:56)
    at node_modules/@octokit/auth-app/dist-web/index.js (version.js:1:14)
    at __init (chunk-AC2VUBZ6.js?v=6002d3fa:8:56)
    at @octokit_auth-app.js?v=6002d3fa:8:1
```

**Comment by wolfy1339** [comment]
Are you building for SSR or a regular browser build?

**Comment by HeavenlyEntity** [comment]
Client side only. 📱

**Comment by wolfy1339** [comment]
Okay, it makes sense that it's trying to use `dist-web` then

The project is supposed to work in a browser. However, to use it in a browser you shouldn't be using the package from `npm`.

This should get fixed once we switch to native ESM modules, and can specify different exports for the browser

See https://github.com/octokit/request.js/issues/456

In the meantime, use `esm.sh` with all Octokit modules.
ex: https://esm.sh/@octokit/plugin-throttling@5.1.1

**Comment by HeavenlyEntity** [comment]
@wolfy1339 duuude, yeah seems the current modules are off for the browser configuration. This is a great start to get working with Octokit `esm.sh` did the trick! You're awesome dude! Thank you! 🎉

**Comment by andrico-anima** [comment]
We're observing the same issue. I'm not sure if it's because we're using the Vite `commonjs` plugin, but adding an alias for the Octokit throttling packages resolves the problem:

```
  resolve: {
    alias: {
      ...stdLibBrowser, fs: 'memfs', '~': path.resolve('.'),
      // There's a bug where the browser version of the package doesn't work when consumed via NPM.
      // https://github.com/octokit/plugin-throttling.js/issues/586#issuecomment-1519166254
      "@octokit/plugin-throttling": "https://esm.sh/@octokit/plugin-throttling@5.1.1"
    }
  },
```

This isn't a great solution for us because:

- Octokit is a dependency of another dependency. 
- It'll be easy for the version of Octokit we've specified in our Vite config to go out of sync with the version installed by our dependency.

I tried making a repro case, but haven't been able to get it failing. If I'm able to, I'll spend a bit more time trying to make something I can share with you both, but I've provided my (hopefully) temporary workaround for others too
